api/ttn/lorawan/v3/devicerepository.proto

```diff
@@ -476,6 +484,11 @@ message MessagePayloadEncoder {
   repeated Example examples = 4 [(validate.rules).repeated.max_items = 20];
 }
 
+message EndDeviceProvisioningInfo {
+  EndDeviceModel end_device_model = 1;
+  EndDeviceTemplate end_device_template = 2;
+}
```

Comment:
1. I think this should be a response message; `GetEndDeviceProvisioningInfoResponse` unless it is an already-defined message. If you define a message just for a request or response, it should be a request or response message. This way we don't introduce messages with common names that are not commonly used except as request/response messages.
2. Isn't the available information more specific than the model? Don't we have `EndDeviceVersionIdentifiers` here?

Reply:
Item 1 is done.

Reply:
For item 2, I checked the network calls made by the console, EndDeviceVersionIdentifiers is the only thing that is needed for creating a device, as the end device ids are provided in the QR or have to be filled in:
```json
{
  "end_device": {
    "ids": {
      "join_eui": "0928745098237059",
      "dev_eui": "9283450982374509",
      "device_id": "dev2",
      "application_ids": {
        "application_id": "app1"
      }
    },
    "version_ids": {
      "brand_id": "the-things-industries",
      "model_id": "generic-node-sensor-edition",
      "hardware_version": "1.1",
      "firmware_version": "1.0",
      "band_id": "EU_863_870"
    },
    "network_server_address": "localhost",
    "application_server_address": "localhost",
    "join_server_address": "localhost"
  },
  "field_mask": {
    "paths": [
      "version_ids.brand_id",
      "version_ids.model_id",
      "version_ids.hardware_version",
      "version_ids.firmware_version",
      "version_ids.band_id",
      "network_server_address",
      "application_server_address",
      "join_server_address"
    ]
  }
}
```

However, if we want to reduce the number of calls done by the console, other details are needed about the device model, to show the model description, photo, product sheet url etc. that are needed in the device registration screen:

<img width="773" alt="Screenshot 2024-08-23 at 10 44 26" src="https://github.com/user-attachments/assets/aa4faed6-f80f-4eec-a995-30ea400a235e">

The full model seems relevant to have in the response. Let me know if I am missing something.

Also, regarding the template, does it make sense to have the full template in the response or is it okay to just have the end device? The message definition for the template looks like this:

message EndDeviceTemplate {
  EndDevice end_device = 1 [(validate.rules).message.required = true];
  google.protobuf.FieldMask field_mask = 2;
  string mapping_key = 3 [(validate.rules).string.max_len = 100];
}

So the EndDeviceTemplate is the EndDevice plus the field_mask and the mapping_key. Given that the GetEndDeviceProvisioningInfoResponse is contains two messages, the end_device_template and the end_device_model, a field_mask only for one of the messages seems unintuitive. What if we do:

message GetEndDeviceProvisioningInfoResponse {
  EndDeviceModel end_device_model = 1;
  EndDevice end_device = 2;
}

And if needed we can add a FieldMask for the full response, instead of having a field_mask for only one of the items? What do you think?

Reply:
@johanstokking what do you think about this?

Reply:
The API is not designed for the Console. If a presentation layer, like the Console, wants to present some additional data, it can request it using (existing) API.

In this case, my comment to return version identifiers is not only to return the most specific data possible for this endpoint, but also because the model does not contain the version identifiers. It contains the entire model, with potentially many firmware versions and regions. So clients cannot use it to create a new device. The goal is to know the version identifiers, so return that.

---

The way `EndDeviceTemplate` works is that it contains the `EndDevice` plus the field mask for the fields that are part of the template.

When setting a device to the NS, AS and JS registries, a field mask is required. So when creating a device, API clients (including the Console), need a field mask.

If we just return `EndDevice` with all kinds of fields set, the API client would have to figure out what fields are set to non-default values, and construct a field mask themselves. That is not API client friendly and may lead to all kinds of problems with default values etc.

So what API clients do, is they get the `EndDevice` plus the field mask of fields that are set. That is the template for creating an end device. Then, the version identifiers are set and added to the field mask. Most likely, the `EndDevice` is complete and the field mask is known, so it can be created.

Reply:
Okay, I understand now. I changed the response to use EndDeviceVersionIdentifiers instead of EndDeviceModel. I left the EndDeviceTemplate as is.